indexer/displacement_manager.hpp

```diff
@@ -0,0 +1,212 @@
+    return m_feature < rhs.m_feature;
+  }
+
+  uint64_t GetCell() const { return UINT64_FROM_UINT32(m_cellHi, m_cellLo); }
+  uint32_t GetFeature() const { return m_feature; }
+private:
+  uint32_t m_cellLo;
+  uint32_t m_cellHi;
+  uint32_t m_feature;
+};
+static_assert(sizeof(CellFeaturePair) == 12, "");
+#ifndef OMIM_OS_LINUX
+static_assert(is_trivially_copyable<CellFeaturePair>::value, "");
+#endif
+
+class CellFeatureBucketTuple
+{
+public:
+  CellFeatureBucketTuple() : m_bucket(0) {}
+  CellFeatureBucketTuple(CellFeaturePair const & p, uint32_t bucket) : m_pair(p), m_bucket(bucket)
+  {
+  }
+
+  bool operator<(CellFeatureBucketTuple const & rhs) const
+  {
+    if (m_bucket != rhs.m_bucket)
+      return m_bucket < rhs.m_bucket;
+    return m_pair < rhs.m_pair;
+  }
+
+  CellFeaturePair const & GetCellFeaturePair() const { return m_pair; }
+  uint32_t GetBucket() const { return m_bucket; }
+private:
+  CellFeaturePair m_pair;
+  uint32_t m_bucket;
+};
+static_assert(sizeof(CellFeatureBucketTuple) == 16, "");
+#ifndef OMIM_OS_LINUX
+static_assert(is_trivially_copyable<CellFeatureBucketTuple>::value, "");
+#endif
+
+/// Displacement manager filters incoming single-point features to simplify runtime
+/// feature visibility displacement.
+template <class TSorter>
+class DisplacementManager
+{
+public:
+  DisplacementManager(TSorter & sorter) : m_sorter(sorter) {}
+
+  /// Add feature at bucket (zoom) to displacable queue if possible. Pass to bucket otherwise.
+  template <class TFeature>
+  void Add(vector<int64_t> const & cells, uint32_t bucket, TFeature const & ft, uint32_t index)
+  {
+    // Add to displacable storage if we need to displace POI.
+    if (bucket != scales::GetUpperScale() && IsDisplaceable(ft))
+    {
+      m_storage.emplace_back(cells, ft, index, bucket);
+      return;
+    }
+
+    // Pass feature to the index otherwise.
+    for (auto const & cell : cells)
+      m_sorter.Add(CellFeatureBucketTuple(CellFeaturePair(cell, index), bucket));
+  }
+
+  /// Check features intersection and supress drawing of intersected features.
+  /// As a result some features may have bigger scale parametr than style describes.
+  /// But every feature has MaxScale at least.
+  /// After all features passed to sorter.
+  void Displace()
+  {
+    m4::Tree<DisplaceableNode> acceptedNodes;
+
+    sort(m_storage.begin(), m_storage.end());
+    for (auto const & node : m_storage)
+    {
+      uint32_t scale = node.m_minScale;
+      for (; scale < scales::GetUpperScale(); ++scale)
+      {
+        float const delta = CalculateDeltaForZoom(scale);
+
+        m2::RectD const displacementRect(node.m_center, node.m_center);
+        bool isDisplaced = false;
+        acceptedNodes.ForEachInRect(m2::Inflate(displacementRect, {delta, delta}),
+            [&isDisplaced, &node, &delta, &scale](DisplaceableNode const & rhs)
+            {
+              if (node.m_center.SquareLength(rhs.m_center) < delta * delta && rhs.m_maxScale > scale)
```

Comment:
BTW, how about to create a pre-computed variable for delta * delta? Like deltaS2?